mybuild_docs/sphinx/man/2.10.0/fpgadiag.8
.\" Man page generated from reStructuredText.
.
.TH "FPGADIAG" "8" "Nov 22, 2023" "2.10.0" "OPAE"
.SH NAME
fpgadiag \- FPGA diagnosis and testing tool
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
fpgadiag [\-m | \-\-mode=] <mode> [\-t | \-\-target=] <target> [options]
.ft P
.fi
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
Includes several tests to diagnose, test, and report on the FPGA hardware.
.sp
\fB<mode>\fP chooses which test to run.
\fB<target>\fP specifies the platform that runs the test.
\fB<target>\fP can be either \fBfpga\fP or \fBase\fP where \fBase\fP\&.
\fB<ase>\fP is the abbreviation for Accelerator Simulation Environment.
.sp
The \fB<mode>\fP selects from the  following tests:
.sp
\fBlpbk1\fP
.sp
This test runs a loopback test on the number of cachelines specified with
the \fBBEGIN\fP option. \fBfpgadiag\fP sets up source and  destination buffers in
main memory. The FPGA then performs a \fBmemcpy\fP from a source buffer to the
destination buffer, one cacheline at a time.
.sp
A cacheline is 64 bytes. When \fBBEGIN = END\fP, the test performs one iteration. When
\fBBEGIN = END + x\fP, the test performs \fBx\fP iterations. The first iteration consists
of copying \fBBEGIN\fP cachelines; the second iteration consists of copying
\fBBEGIN+1\fP cache lines. The third iteration consists of copying \fBBEGIN+2\fP
cache lines, and so on.
.sp
The latency is shown as the number of clock cycles.
.sp
When you specify \fBMULTI\-CL\fP, you copy \fBMULTI\-CL\fP cache lines at a time.
The WR\-FENCE chooses on which virtual channel the WrFence occurs.
.sp
If you specify continuous mode with \fB\-\-cont\fP, the program iterates
until the timeout specified in \fBTIMEOUT\fP completes.
.sp
\fBread\fP
.sp
This test performs reads. Use this test to measure read bandwidth.
.sp
\fBwrite\fP
.sp
This test performs writes. Use it to measure write bandwidth.
.sp
\fBtrput\fP
.sp
This test measures both read and write bandwidth by performing 50% read and
50% write tests.
.sp
\fBsw\fP
.sp
This is a send\-and\-respond (ping\-pong) test. One side sends data and
waits for response.
.sp
Each test requires a particular AF. Before running a test,
make sure the required AF is properly configured
on the platform.
.INDENT 0.0
.IP \(bu 2
The lpbk1 test requires the nlb mode 0 AF.
.IP \(bu 2
The trput test requires the nlb mode 3 AF.
.IP \(bu 2
The sw test requires the nlb mode 7 AF. This AF is only available for the integrated FPGA platform.
You cannot run it on the PCIe accelerator card (PAC).
.UNINDENT
.sp
\fBfpgalpbk\fP
.sp
This enable/disable FPGA loopback.
.sp
\fBfpgastats\fP
.sp
This get fpga mac statistics.
.sp
\fBmactest\fP
.sp
This compare mac addresses that read from MAC ROM with mac addresses read from Host side.
.SH OPTIONS
.SS Common options
.sp
\fB\-\-help, \-h\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Print help information and exit.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-target=, \-t\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
This switch specifies fpga (hardware) or ase (simulation). The default=fpga.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-mode=, \-m\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
The test to run. The valid values are \(galpbk1\(ga, \(garead\(ga,
\(gawrite\(ga, \(gatrput\(ga, and \(gasw\(ga.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-config=, \-c\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
A configuration file in the JSON format that specifies options for a test.
If an option is specified both in the configuration file and on the command 
line, the value in the configuration file takes precedence.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-dsm\-timeout\-usec\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Timeout in microseconds for test completion. The test fails if not completed by 
specified timeout. The default=1000000.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-socket\-id=, \-s\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Socket ID encoded in FPGA Interface Manager (FIM). The default=0. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-bus=, \-B\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Bus number of the PCIe device. The default=0. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-device=, \-D\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Device number of the PCIe device. The default=0. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-function=, \-F\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Function number of the PCIe device. The default=0. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-freq=, \-T\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Clock frequency (in Hz) used for bandwidth calculation. The default=400000000 Hz (400 MHz). 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
This frequency is used only when the software cannot infer the frequency from the accelerator.
.UNINDENT
.UNINDENT
.sp
\fB\-\-suppress\-hdr, \-S\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Suppress column headers for text output. The default=off.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-csv, \-V\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Comma separated value format. The default=off. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-suppress\-stats\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Suppress statistics output at the end of test. The default=off.
.ft P
.fi
.UNINDENT
.UNINDENT
.SS \fBlpbk1\fP test options
.sp
\fB\-\-guid=, \-g\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
AFU ID to enumerate. The default=D8424DC4\-A4A3\-C413\-F89E\-433683F9040B. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-begin=B, \-b\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
1 <= B <= 65535. The default=1, B = number of cache lines. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-end=E, \-e\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
1 <= E <= 65535. The default=B, B and E designate number of cache lines. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-multi\-cl=M, \-u\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
M can equal 1, 2, or 4. The default=1. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-cont, \-L\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Continuous mode. The default=off. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-timeout\-usec=, \-\-timeout\-msec=, \-\-timeout\-sec=, \-\-timeout\-min=, \-\-timeout\-hour=\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
timeout for \-\-cont mode. The default for all options is 0. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-cache\-policy=, \-p\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be wrline\-I, wrline\-M, or wrpush\-I The default=wrline\-M.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-cache\-hint=, \-i\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be rdline\-I or rdline\-S. The default=rdline\-I.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-read\-vc=, \-r\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be auto, vl0, vh0, vh1, random. The default=auto. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-write\-vc=, \-w\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be auto, vl0, vh0, vh1, random. The default=auto. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-wrfence\-vc=, \-f\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be auto, vl0, vh0, vh1. The default=auto. 
.ft P
.fi
.UNINDENT
.UNINDENT
.SS \fBread\fP test options
.sp
\fB\-\-guid=, \-g\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
AFU ID to enumerate. The default=F7DF405C\-BD7A\-CF72\-22F1\-44B0B93ACD18. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-begin=B, \-b\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
1 <= B <= 65535. The default=1, B = number of cache lines. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-end=E, \-e\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
1 <= E <= 65535. The default=B, B and E designate number of cache lines. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-multi\-cl=M, \-u\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
M can equal 1, 2, or 4. The default=1. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-strided\-access=S, \-a\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
1<= S <= 64. The default=1. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-cont, \-L\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Continuous mode. The default=off. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-timeout\-usec=, \-\-timeout\-msec=, \-\-timeout\-sec=, \-\-timeout\-min=, \-\-timeout\-hour=\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
timeout for \-\-cont mode. The default for all options is 0.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-cache\-hint=, \-i\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be rdline\-I or rdline\-S. The default=rdline\-I. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-warm\-fpga\-cache \-H; \-\-cool\-fpga\-cache \-M\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Try to prime the cache with hits. The default=off. Try to prime the 
cache with misses. The default=off.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-cool\-cpu\-cache, \-C\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Try to prime the cpu cache with misses. The default=off. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-read\-vc=, \-r\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be auto, vl0, vh0, vh1, random. The default=auto 
.ft P
.fi
.UNINDENT
.UNINDENT
.SS \fBwrite\fP test options
.sp
\fB\-\-guid=, \-g\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
AFU ID to enumerate. The default=F7DF405C\-BD7A\-CF72\-22F1\-44B0B93ACD18 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-begin=B, \-b\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
1 <= E <= 65535. The default=B, B and E designate number of cache lines. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-multi\-cl=M, \-u\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
M can equal 1, 2, or 4. The default=1.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-strided\-access=S, \-a\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
1<= S <= 64. The default=1.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-cont, \-L\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Continuous mode. The default=off.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-timeout\-usec=, \-\-timeout\-msec=, \-\-timeout\-sec=, \-\-timeout\-min=, \-\-timeout\-hour=\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
timeout for \-\-cont mode. The default for all options is 0.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-cache\-policy=, \-p\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be wrline\-I, wrline\-M, or wrpush\-I The default=wrline\-M 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-warm\-fpga\-cache \-H; \-\-cool\-fpga\-cache \-M\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Try to prime the cache with hits. The default=off. Try to prime the 
cache with misses. The default=off. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-cool\-cpu\-cache, \-C\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Try to prime the cpu cache with misses. The default=off. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-write\-vc=, \-w\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be auto, vl0, vh0, vh1, random. The default=auto. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-wrfence\-vc=, \-f\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be auto, vl0, vh0, vh1, random. The default=\(gaWRITE\-VC\(ga.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-alt\-wr\-pattern, \-l\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Alternate Write Pattern. The default=off. 
.ft P
.fi
.UNINDENT
.UNINDENT
.SS \fBtrput\fP test options
.sp
\fB\-\-guid=, \-g\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
AFU ID to enumerate. The default=F7DF405C\-BD7A\-CF72\-22F1\-44B0B93ACD18.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-begin=B, \-b\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
1 <= B <= 65535. The default=1, B = number of cache lines. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-end=E, \-e\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
1 <= E <= 65535. The default=B, B and E designate number of cache lines. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-multi\-cl=M, \-u\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
M can equal 1, 2, or 4. The default=1. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-strided\-access=S, \-a\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
1<= S <= 64. The default=1 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-cont, \-L\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Continuous mode. The default=off. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-timeout\-usec=, \-\-timeout\-msec=, \-\-timeout\-sec=, \-\-timeout\-min=, \-\-timeout\-hour=\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
timeout for \-\-cont mode. The default for all options is 0.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-cache\-policy=, \-p\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be wrline\-I, wrline\-M, or wrpush\-I The default=wrline\-M. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-cache\-hint=, \-i\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be rdline\-I or rdline\-S. The default=rdline\-I. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-read\-vc=, \-r\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be auto, vl0, vh0, vh1, random. The default=auto. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-write\-vc=, \-w\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be auto, vl0, vh0, vh1, random. The default=auto. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-wrfence\-vc=, \-f\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be  auto, vl0, vh0, vh1. The default=\(gaWRITE\-VC\(ga.
.ft P
.fi
.UNINDENT
.UNINDENT
.SS \fBsw\fP test options
.sp
\fB\-\-guid=, \-g\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
AFU ID to enumerate. The default=7BAF4DEA\-A57C\-E91E\-168A\-455D9BDA88A3. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-begin=B, \-b\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
1 <= B <= 65535. The default=1, B = number of cache lines. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-end=E, \-e\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
1 <= E <= 65535. The default=B, B and E designate number of cache lines. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-cache\-policy=, \-p\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be wrline\-I, wrline\-M, or wrpush\-I. The default=wrline\-M. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-cache\-hint= \-i\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be rdline\-I or rdline\-S. The default=rdline\-I. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-read\-vc=, \-r\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be auto, vl0, vh0, vh1, random The default=auto. 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-write\-vc=, \-w\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be auto, vl0, vh0, vh1, random The default=auto.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-wrfence\-vc=, \-f\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be auto, vl0, vh0, vh1. The default=\(gaWRITE\-VC\(ga.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-notice=, \-N\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be poll or csr\-write. The default=poll. 
.ft P
.fi
.UNINDENT
.UNINDENT
.SS \fBEnable FPGA N3000 Ethernet group VFIO mdev\fP
.sp
FPGA DFL driver does not support any ioctls to read/write ethernet group info and registers.
Users can read/write eth group registers by enabling VFIO mdev. Unbind the dfl_eth_group driver and bind vfio\-mdev\-dfl
driver for ethernet group dfl\-device; then userspace can take full control of ethernet group feature id 10.
.sp
Ethernet group must be enabled before running fpgalpbk, mactest tools.
.SS \fBSteps to enable/create vfio mdev\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
unbind eth group feature id 10:
    echo dfl\-fme.0.8 > /sys/bus/dfl/drivers/dfl\-eth\-group/unbind
    echo dfl\-fme.0.7 > /sys/bus/dfl/drivers/dfl\-eth\-group/unbind
bind to vfio\-mdev\-dfl:
    echo vfio\-mdev\-dfl > /sys/bus/dfl/devices/dfl\-fme.0.7/driver_override
    echo vfio\-mdev\-dfl > /sys/bus/dfl/devices/dfl\-fme.0.8/driver_override
load vfio driver:
    modprobe vfio_pci
    modprobe vfio_iommu_type1
    modprobe vfio_mdev
    modprobe vfio_mdev_dfl
trigger mdev:
    echo dfl\-fme.0.7 >/sys/bus/dfl/drivers_probe
    echo dfl\-fme.0.8 >/sys/bus/dfl/drivers_probe
    echo 83b8f4f2\-509f\-382f\-3c1e\-e6bfe0fa1001 > /sys/bus/dfl/devices/dfl\-fme.0.7/mdev_supported_types/vfio\-mdev\-dfl\-1/create
    echo 83b8f4f2\-509f\-382f\-3c1e\-e6bfe0fa1002 > /sys/bus/dfl/devices/dfl\-fme.0.8/mdev_supported_types/vfio\-mdev\-dfl\-1/create

linux kerenl msg after enabling mdev:
    i40e 0000:b3:00.0 eth1: NIC Link is Down
    i40e 0000:b1:00.1 eth0: NIC Link is Down
    vfio\-mdev\-dfl dfl\-fme.2.7: MDEV: Registered
    vfio\-mdev\-dfl dfl\-fme.2.8: MDEV: Registered
    vfio_mdev 83b8f4f2\-509f\-382f\-3c1e\-e6bfe0fa1005: Adding to iommu group 140
    vfio_mdev 83b8f4f2\-509f\-382f\-3c1e\-e6bfe0fa1005: MDEV: group_id = 140
    vfio_mdev 83b8f4f2\-509f\-382f\-3c1e\-e6bfe0fa1006: Adding to iommu group 141
    vfio_mdev 83b8f4f2\-509f\-382f\-3c1e\-e6bfe0fa1006: MDEV: group_id = 141
.ft P
.fi
.UNINDENT
.UNINDENT
.SS \fBRemove vfio mdev\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
    echo 1 | sudo tee /sys/bus/mdev/devices/83b8f4f2\-509f\-382f\-3c1e\-e6bfe0fa1002/remove
    echo 1 | sudo tee /sys/bus/mdev/devices/83b8f4f2\-509f\-382f\-3c1e\-e6bfe0fa1001/remove

    rmmod vfio_mdev_dfl
    modprobe dfl_eth_group

    echo dfl\-fme.0.7 >/sys/bus/dfl/drivers_probe
    echo dfl\-fme.0.8 >/sys/bus/dfl/drivers_probe

    echo dfl\-eth\-group > /sys/bus/dfl/devices/dfl\-fme.0.7/driver_override
    echo dfl\-eth\-group > /sys/bus/dfl/devices/dfl\-fme.0.8/driver_override
.ft P
.fi
.UNINDENT
.UNINDENT
.SS \fBfpgalpbk\fP test options
.sp
\fB\-\-enable\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Enable fpga phy loopback.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-disable\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Disable fpga phy loopback.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-direction\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be local, remote.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-type\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be serial, precdr, postcdr.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-side\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Can be line, host.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-port\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
0 <= port <= 7, the default is all.
.ft P
.fi
.UNINDENT
.UNINDENT
.SS \fBmactest\fP test options
.sp
\fB\-\-offset\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Read mac addresses from an offset, The default=0.
.ft P
.fi
.UNINDENT
.UNINDENT
.SH EXAMPLES
.sp
This command starts a \fBlpbk1\fP test for the FPGA on bus \fB0x5e\fP\&. The test
copies 57535, 57536, 57537 … up to 65535 cache lines, one line at a time.
The test prints output in the comma separated values (CSV) format with the
header suppressed.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\&./fpgadiag \-\-mode=lpbk1 \-\-target=fpga \-V \-\-bus=0x5e \-\-begin=57535
\-\-end=65535 \-\-cache\-hint=rdline\-I \-\-cache\-policy=wrpush\-I \-\-multi\-cl=1
\-\-write\-vc=vl0 \-\-read\-vc=vh1 \-\-wrfence\-vc=auto
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This command starts a \fBread\fP test on the FPGA located on bus \fB0xbe\fP\&. The test
reads 2045 cache lines in the continuous mode with a 15\-second timeout period.
The reads use a strided pattern with a 10\-byte stride length.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\&./fpgadiag \-\-mode=read \-\-target=fpga \-V \-\-bus=0xbe \-\-begin=2045 \-\-cont
\-\-timeout\-sec=15 \-\-cache\-hint=rdline\-I \-\-multi\-cl=1 \-a=10 
\-\-read\-vc=auto \-\-wrfence\-vc=auto
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This command starts a \fBsw\fP test on the FPGA located on bus \fB0xbe\fP\&. The test
signals completion using a CSR write.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\&./fpgadiag \-\-mode=sw \-\-target=fpga \-V \-\-bus=0xbe \-\-begin=4 \-\-end=8192
\-\-cache\-hint=rdline\-I \-\-cache\-policy=wrline\-I \-\-notice=csr\-write \-\-write\-vc=vl0
\-\-wrfence\-vc=auto \-\-read\-vc=random 
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This command enable a \fBfpgalpbk\fP on the FPGA located on bus \fB0xbe\fP\&.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\&./fpgadiag \-m fpgalpbk \-\-bus 0xbe \-\-enable \-\-direction local \-\-type postcdr
\-\-side host
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This command show \fBfpgastats\fP on the FPGA located on bus \fB0xbe\fP\&.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\&./fpgadiag \-m fpgastats \-\-bus 0xbe
.ft P
.fi
.UNINDENT
.UNINDENT
.SH TROUBLESHOOTING
.sp
When a test fails to run or gives errors, check the following:
.INDENT 0.0
.IP \(bu 2
Is the Intel FPGA driver properly installed?
See \fI\%Installation Guide\fP
for driver installation instructions.
.IP \(bu 2

binaries/fpgadiag/opae/diag/fpgadiag.py
#! /usr/bin/env python3
# Copyright(c) 2017-2023 Intel Corporation
#
# Redistribution  and  use  in source  and  binary  forms,  with  or  without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of  source code  must retain the  above copyright notice,
# this list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation
# and/or other materials provided with the distribution.
# * Neither the name  of Intel Corporation  nor the names of its contributors
# may be used to  endorse or promote  products derived  from this  software
# without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
# IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
# LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
# CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
# SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
# INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
# CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import os
import argparse
import subprocess
import sys
from subprocess import CalledProcessError

cwd = os.path.dirname(sys.argv[0])
cmd_list = ['lpbk1', 'read', 'write', 'trput', 'sw',
            'fvlbypass', 'fpgalpbk', 'mactest', 'fpgastats', 'fpgamac']
cmd_map = {'lpbk1': ['nlb0'],
           'read': ['nlb3', '--mode=read'],
           'write': ['nlb3', '--mode=write'],
           'trput': ['nlb3', '--mode=trput'],
           'sw': ['nlb7'],
           'fvlbypass': ['fvlbypass'],
           'fpgalpbk': ['fpgalpbk'],
           'mactest': ['mactest'],
           'fpgastats': ['fpgastats'],
           'fpgamac': ['fpgamac']}


def main():
    parser = argparse.ArgumentParser(add_help=False)
    parser.add_argument('-t', '--target',
                        default='fpga',
                        choices=['fpga', 'ase'],
                        help='choose target. Default: fpga')
    parser.add_argument('-m', '--mode', choices=cmd_list,
                        help=('choose test mode. Combine this with'
                              '-h or --help to see detail help message'
                              'for each mode.'))

    args, leftover = parser.parse_known_args()

    if args.mode:
        cmdline = cmd_map[args.mode] + leftover
    else:
        parser.print_help()
        exit(1)

    cmdline[0] = os.path.join(cwd, cmdline[0])
    cmdline = cmdline + ['-t', args.target] + leftover

    try:
        subprocess.run(cmdline, check=True)
    except CalledProcessError as e:
        exit(e.returncode)


if __name__ == "__main__":
    main()
